# Assignment 5: Association mining

## Objective of this assignment
The overall objective is to understand how frequent itemsets can be extracted by
the Apriori algorithm and be able to calculate and interpret association rules in terms of support and confidence.

## ** Important: ** When handing in your homework:
+ Hand in the notebook (and nothing else) named as follows: StudentName1_snumber_StudentName2_snumber.ipynb
+ Provide clear and complete answers to the questions below under a separate header (not hidden somewhere in your source code), and make sure to explain your answers / motivate your choices. Add Markdown cells where necessary.
+ Source code, output graphs, derivations, etc., should be included in the notebook.
+ Hand-in: upload to Blackboard.
+ Include name, student number, assignment (especially in filenames)!
+ When working in pairs only one of you should upload the assignment, and report the name of your partner in your filename.
+ For problems or questions: use the BB discussion board or email the student assistants.


## Advised Reading and Exercise Material
**The following reading material is recommended:**

- Pang-Ning Tan, Michael Steinbach, and Vipin Kumar, *Introduction to Data Mining*, section 6.


## Additional Tools
For this exercise you will need to load the provided *apriorimining.py* script. 


##  5.1 Association mining for course data 
We will use the Apriori algorithm to automatically mine for associations. The Apriori algorithm is an adapted version of the script found here: https://github.com/nalinaksh/Association-Rule-Mining-Python

Check out the script and doc and check if you understand how the association rules are computed. 



In [2]:
import Toolbox.apriorimining


#### 5.1.1

(0 points) Look at the data file `Data/courses.txt` into Python. The data is represented in Table 1. Inspect the file Data/courses.txt and make sure you understand how the data in Table 1 is stored in the text file.

##### Table 1
|#  |   History |Math| Biology| Spanish | Economics| Physics | Chemistry | English  |  
| :-------------: |:-------------:| :-----------:| :----------:| :------------:|:-------------:| :------------:|  :-------------: | :-------------: |
|student 1 | 0| 1 | 0 | 0 | 1| 1 |1 |1   
|student 2 | 1| 1 | 1 | 0 | 0| 1 |1 |1   
|student 3 | 0| 1 | 0 | 1 | 0| 1 |0 |1   
|student 4 | 0| 0 | 1 | 0 | 0| 1 |1 |0   
|student 5 | 0| 1 | 0 | 0 | 0| 1 |1 |0        
|student 6 | 0| 1 | 1 | 0 | 0| 1 |1 |1   


#### 5.1.2
(1 point) We will analyze the data in Table 1 automatically using the function `apriorimining.generate_association_rules()` from the script. Analyze the data with $ minsupport  \geq 80 \% $ and $ minconfidence \geq 100 \%$.What
are the generated association rules? What kind of conclusions can you make based on these association rules about the subjects that students took?  
  


In [3]:
import itertools
from Toolbox import apriorimining
with open('Data/courses.txt') as f:
    content = f.readlines()
content = [x.strip() for x in content] 

apriorimining.generate_association_rules()

Please enter support value in %: 80
Please enter confidence value in %: 100
Enter the max number of rules you want to see (enter 0 to see all rules): 0
Please enter filepath\filename (for Windows), or filepath/filename (for UNIX/Mac), and extension: Data/courses.txt
---------------TOP 10 FREQUENT 1-ITEMSET-------------------------
set= { 6 },  sup= 100.0
set= { 2 },  sup= 83.33
set= { 7 },  sup= 83.33
-----------------------------------------------------------------
-------TOP 10 (or less) FREQUENT 2-ITEMSET------------------------
set= { 2, 6 },  sup= 83.33
set= { 6, 7 },  sup= 83.33
------------------------------------------------------------------
---------------------ASSOCIATION RULES------------------
--------------------------------------------------------
Rule #1: {  } ==> { 6 }, sup= 100.00, conf= 100.00

Rule #2: { 2 } ==> { 6 }, sup= 83.33, conf= 100.00

Rule #3: { 7 } ==> { 6 }, sup= 83.33, conf= 100.00
--------------------------------------------------------


*Double click to type your answer for 5.1.2 here:*
In Rule #1, regradless on course student take, evry student choose 6(Pysics).
In Rule #2, the percentage of student who choose 2(Math) and 6(Pysics) is 83.33(5/6) and if one student chose 2(Math), he also choose 6(Pysics)
In Rule #3, the percentage of student who choose 7(Chemistry) and 6(Pysics) is 83.33(5/6) and if one student chose 7(Chemistry), he also choose 6(Pysics)


  ##  5.2 Association mining for MovieLens data 
  
  
  In this part of the exercise we consider a Market Basket data set containing 943 users purchases of 1682 movies. A total of 100,000 movies
have been purchased.The data set is called MovieLens100K and is provided by http://www.grouplens.org/node/73, see also the readme `MovieLensData.txt` in the data folder. The data currently considered is not the original data but modified for the apriori algorithm.

#### 5.2.1
  (0 points) The MovieLens data is stored in the file MovieLensData.txt. Inspect the file to see how the data is stored.


#### 5.2.2 
  (1 point) Find association rules using the function below with $ minsupport  \geq 30 \% $ and $ minconfidence \geq 80 \%$. What are the associations with strongest confidence? Do these associations make sense? The script can use file Data/u.item to print the movie titles in stead of numbers. If you enter filename `MovieLensData.txt`, the script will provide an additional option for this. 
  

In [4]:
'''
C1 = {}
transactions = 0
D = []
T = []
with open("Data/MovieLensData.txt", "r") as f:
    for line in f:
        T = []
        transactions += 1
        for word in line.split():
            T.append(word)
            if word not in C1.keys():
                C1[word] = 1
            else:
                count = C1[word]
                C1[word] = count + 1
        D.append(T)
print(D[0])
'''

apriorimining.generate_association_rules()

Please enter support value in %: 30
Please enter confidence value in %: 80
Enter the max number of rules you want to see (enter 0 to see all rules): 0
Please enter filepath\filename (for Windows), or filepath/filename (for UNIX/Mac), and extension: Data/MovieLensData.txt
Do you want to print sets and rules with Movie names in stead of numbers? [y/n]: y
---------------TOP 10 FREQUENT 1-ITEMSET-------------------------
set= { Star Wars (1977) },  sup= 61.82
set= { Contact (1997) },  sup= 53.98
set= { Fargo (1996) },  sup= 53.87
set= { Return of the Jedi (1983) },  sup= 53.76
set= { Liar Liar (1997) },  sup= 51.43
set= { English Patient, The (1996) },  sup= 51.01
set= { Scream (1996) },  sup= 50.69
set= { Toy Story (1995) },  sup= 47.93
set= { Air Force One (1997) },  sup= 45.71
set= { Independence Day (ID4) (1996) },  sup= 45.49
-----------------------------------------------------------------
-------TOP 10 (or less) FREQUENT 2-ITEMSET------------------------
set= { Return of the Jedi (1

The rule which has highst confidence(99.66) is #75{ Empire Strikes Back, Raiders of the Lost Ark, Return of the Jedi  } ==> { Star Wars }, which occupies 31.18%. It makes sense because all of the three movies are 'Star Wars' series.


#### 5.2.3 
(1 point) Which movies have been watched by the most users? There are only few rules with more than three items. Why?

number 50 is watched most, 60% of people watched it. one of the reason why there are few rules with more than 3 item is there is just one set for TOP 10 (or less) FREQUENT 4-ITEMSET. In other words, there is a few people who share 3 movies. The other reason is there are only 3 popular 'Star Wars' series, so it is hard to find rules more than three items 

#### 5.2.4
(0.5 points) Often we are interested in rules with high confidence. Is it possible for
itemsets to have very low support but still have a very high confidence?

$$\text{The rule of form } \{a\rightarrow b\} \text{ might have a low support, but if most sets that contain a contain b then }  c(a \rightarrow b) \text{ will be high.} $$


## 5.3 Calculating support, confidence and interest

Calculate these measures and write down how you computed things, not just the answers. 


#### 5.3.1
 Suppose we have market basket data consisting of 100 transactions and 20 items. The support for item $ \text{a} = 45 \%$, the support for item $ \text{b} = 80 \%$ and the support for itemset $ \text{ {a,b }} = 30 \%$. Let the support and confidence thresholds be 20$ \%$ and 60$ \%$, respectively.
  
1. (0.5 points) Compute the confidence of the association rule $ \text{ {a } } \rightarrow   \text{{b }} $. Is the rule interesting according to the confidence measure?

2. (0.5 points) Compute the interest measure (or lift, see slide 44 of chapter 6) for the association pattern $ \text{ {a,b}}$. Describe the nature of the relationship between item $ \text{a}$ and item $ \text{b}$  in terms of the itemset measure.
3. (1 points) What conclusion can you draw from the results of parts (1) and (2)?

$$1.)\quad c(a \rightarrow b)=\frac{\sigma(a,b)}{\sigma(a)}=\frac{30%}{45%}=\frac{2}{3}=0.\overline6
\\
\\
\\
2.)\quad Lift = \frac{c(a \rightarrow b)}{s(b)} = \frac{\frac{2}{3}}{\frac{4}{5}}=\frac{5}{6}=0.8\overline3
\\
\text{0.83 < 1  so they're negatively associated}
\\
3.)\quad \text{Both of them are over the required confidance level,}\\
\text{so both of them are significant, so the itemset \{a,b\} is significant.}
$$



#### 5.3.2

1. (1 points) Let $c_1$, $c_2$, and $c_3$ be the confidence values of the rules $ \text{ {a } } \rightarrow   \text{{b }} $, $ \text{ {a } } \rightarrow   \text{{b,c }} $, and $ \text{ {a,c } } \rightarrow   \text{{b }} $ respectively. If we assume that $c_1$, $c_2$, and $c_3$ have different values, what are the possible inequality relationships (e.g. $c_1 \leq c_2 \leq c_3$) among $c_1$, $c_2$, and $c_3$? Which rule has the lowest confidence?
2. (0.5 points) Suppose the confidence of the rules  $ \text{ {a } } \rightarrow   \text{{b }} $ and $ \text{ {b } } \rightarrow   \text{{c }} $ are larger than the confidence threshold. Is it possible that $ \text{ {a } } \rightarrow   \text{{c }} $ has a confidence below that threshold? If no, explain why. If yes, give an example. 

$$ \text{Because the rule } \{a\}\rightarrow \{b,c\} \text { has stronger restrictions than }  \{a\}\rightarrow \{b\} , c_1 \geq c_2 (1)
\\
c(c_2)=\frac{\sigma(a,b,c)}{\sigma(a)}\\
c(c_3)=\frac{\sigma(a,b,c)}{\sigma(a,c)}\\
\sigma(a,c) \leq \sigma(a) \\
\text{ So } c_3 \geq c_2 (2)\\
\text{similarely to }c_2\text{ } c_3 \leq c_1 \text{,because }
\sigma(a,b,c) \leq \sigma(a,b) \Rightarrow c_1 \geq c_3 (3)\\
\text{From (1),(2)&(3)}\Rightarrow c_1 \geq c_3 \geq c_2\\
2.)\quad
\text{It can be lower than the threshod if we only have \{a,b\} and \{b,c\} }\\\text{pairs  in majority because, if we have only}\\\text{ those(in an equal number for these exact confidances) the confidance level for:}\\ c(\{a\}\rightarrow\{b\}) = 1\text{ and }c(\{b\}\rightarrow\{c\}) = 0.5\text{ but }c(\{a\}\rightarrow\{c\}) = 0
$$



#### 5.3.3

(3 points) Consider the relationships between customers who buy high-definition televisions and exercise machines as shown in Table 2 and 3.

1. Compute the odd ratios for both tables.
2. Compute the $\phi$-coefficient for both tables.
3. Compute the interest (or lift, in the book) factor for both tables.

For Table 3 you should compute measures given above separately for College
Students and for Adults. For each of the measures, describe how the direction
of association changes when data is pooled together (Table 2) instead of being
separated into two groups (Table 3)

##### Table 2: Two way contingency table between the sale of high-definition television and exercise machine
| |   Buy Exercise machine |     |     |
| :------------- | -------------:| :-----------:| :----------:| 
| **Buy HDTV     ** | yes | no | total |
| yes  | 105| 87 | 192 | 
| no | 40| 62 | 102 |   
| total | 145 | 149 | 294 | 
 

##### Table 3: Example of three-way contingency table
| | |   Buy Exercise machine |     |     |
|--- | :------------- | -------------:| :-----------:| :----------:| 
|**Customer group** | **Buy HDTV     ** | yes | no | total |
|College students | yes  | 2| 9 | 11 | 
| | no | 5| 20 | 25 |
| Working adults | yes  | 103| 78 | 181 | 
| | no | 35| 42 | 77 |  



$$\text{Table 2:}\\
1.)\quad \text{Odds ratio}=\frac{P(A,B)P(\overline A,\overline B)}{P(A,\overline B)P(\overline A,B)}=\frac{\frac{105}{294}*\frac{62}{294}}{\frac{40}{294}*\frac{87}{294}}=1.8706896551724\\
2.)\quad\Phi-coefficient=\frac{P(A,B)-P(A)P(B)}{\sqrt{P(A)P(B)(1-P(A))(1-P(B))}}=\\= \frac{\frac{105}{294}-\frac{145}{294}*\frac{192}{294}}{\sqrt{\frac{145}{294}*\frac{192}{294}(1-\frac{145}{294})(1-\frac{192}{294})}}=\frac{101\sqrt{\frac{5}{146914}}}{4}\approx 0.147304\\
3.)\quad\text{Interest}=\frac{P(A,B)}{P(A)P(B)}=\frac{\frac{105}{294}}{\frac{145}{294}*\frac{192}{294}}=1.1088362068965\\
\text{Table 3(College students):}\\
1.)\quad \text{Odds ratio}=\frac{P(A,B)P(\overline A,\overline B)}{P(A,\overline B)P(\overline A,B)}=\frac{\frac{2}{33}*\frac{20}{33}}{\frac{5}{33}*\frac{9}{33}}=0.\overline8\\
2.)\quad\Phi-coefficient=\frac{P(A,B)-P(A)P(B)}{\sqrt{P(A)P(B)(1-P(A))(1-P(B))}}=\\= \frac{\frac{2}{33}-\frac{7}{33}*\frac{11}{33}}{\sqrt{\frac{7}{33}*\frac{11}{33}(1-\frac{7}{33})(1-\frac{11}{33})}}=-\frac{1}{2\sqrt{91}}\approx-0.05241424\\
3.)\quad\text{Interest}=\frac{P(A,B)}{P(A)P(B)}=\frac{\frac{2}{33}}{\frac{7}{33}*\frac{11}{33}}=0.\overline{857142}\\
\text{Table 3(Working adults):}\\
1.)\quad \text{Odds ratio}=\frac{P(A,B)P(\overline A,\overline B)}{P(A,\overline B)P(\overline A,B)}=\frac{\frac{103}{258}*\frac{42}{258}}{\frac{35}{258}*\frac{78}{258}}=1.5\overline{846153}\\
2.)\quad\Phi-coefficient=\frac{P(A,B)-P(A)P(B)}{\sqrt{P(A)P(B)(1-P(A))(1-P(B))}}=\\= \frac{\frac{103}{258}-\frac{138}{258}*\frac{181}{258}}{\sqrt{\frac{138}{258}*\frac{181}{258}(1-\frac{138}{258})(1-\frac{181}{258})}}=19\sqrt{\frac{7}{228965}}=0.105055325
3.)\quad\text{Interest}=\frac{P(A,B)}{P(A)P(B)}=\frac{\frac{103}{258}}{\frac{138}{258}*\frac{181}{258}}=1.06389622\\
\text{For all of them the mesures are the highes when only looking at adults}\\\text{ lowest when only looking at students and in the middle when looking at both}
$$